xdsl/ir.py

```diff
@@ -520,12 +520,19 @@ def create(cls: type[OpT],
         return cast(OpT, op)
 
     @classmethod
-    def build(cls: type[OpT],
-              operands: list[Any] | None = None,
-              result_types: list[Any] | None = None,
-              attributes: dict[str, Any] | None = None,
-              successors: list[Any] | None = None,
-              regions: list[Any] | None = None) -> OpT:
+    def build(
+        cls: type[OpT],
+        operands: list[SSAValue | Operation
+                       | list[SSAValue | Operation]]
+        | list[SSAValue | Operation] | list[SSAValue | list[SSAValue]]
+        | None = None,
+        result_types: list[Attribute | list[Attribute]] | list[list[Attribute]]
+        | list[Attribute]
+        | None = None,
+        attributes: dict[str, Attribute] | None = None,
+        successors: list[Block] | None = None,
+        regions: list[Region | list[Operation] | list[Block]] | None = None
+    ) -> OpT:
```

Comment:
I get why you write this, but I don't like reading it..
Do you think we can factor those somehow? Maybe in a TypeAlias?
It would be good for this function's readability, and even reusable in dialects' constructors

Reply:
Same questions for the other complicated signatures!

Reply:
I think they're useful for surfacing the complexity of the type. Hiding them in a typealias feels like sweeping the issue under the rug. I'd rather change the interface to be simpler to understand. One big improvement would come from tuples, since they're covariant, and would let us just write `t[A | t[A]]` instead of `l[A] | l[A | l[A]] | l[l[A]]`.

Reply:
I get your point, and I like this idea. Is there a catch? Maybe we should do just that?

Reply:
Actually, could we use instead `Sequence` instead? It is covariant.

My understanding is that we don't need these things to be lists or tuple anyway, we just need to have a sequence of them.